### PR TITLE
DATE_FORMAT function

### PR DESCRIPTION
**Pull request by lyndonb-bq**
*Issue #710 

*Description of changes:*
Added implementation, unit tests, manual integration tests, and documentation for date_format function.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by chloe-zh — APPROVED** [review]
LGTM!

**Review by rupal-bq — APPROVED** [review]

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/datetime/DateTimeFormatterUtil.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+  private DateTimeFormatterUtil() {
+  }
+
+  /**
+   * Format the date using the date format String.
+   * @param dateExpr the date ExprValue of Date/Datetime/Timestamp/String type.
+   * @param formatExpr the format ExprValue of String type.
+   * @return Date formatted using format and returned as a String.
+   */
+  static ExprValue getFormattedDate(ExprValue dateExpr, ExprValue formatExpr) {
```
Comment (penghuo) [inline core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/datetime/DateTimeFormatterUtil.java:87]: Not sure the use case. Do you try to do something like this format.replace("%\\W", () -> replace).
If so, i recommended to consider the https://docs.oracle.com/javase/7/docs/api/java/util/regex/Matcher.html#appendReplacement(java.lang.StringBuffer,%20java.lang.String)
which provide the more clean code structure.

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/datetime/DateTimeFormatterUtil.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    }
+    return SUFFIX_CONVERTER.getOrDefault(val % 10, SUFFIX_SPECIAL_TH);
+  }
+
+  /**
+   * Goes through format String and replaces any %x with 'x' where x is unmapped.
+   * @param format Incoming format String without mapping completed.
+   * @return Outgoing format String with mapping completed.
+   */
+  private static String literalReplace(String format) {
```
Comment (penghuo) [inline core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/datetime/DateTimeFormatterUtil.java:121]: What is the expected result if the user input the format string which we doesn't support?

Comment (lyndonb-bq) [reply core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/datetime/DateTimeFormatterUtil.java:121]: Do you mean a format we don't support, such as %z? 
According to the documentation for MySQL it should come out as just z, so using the DateTimeFormatter specification, to get a literal out, we wrap in ', so %z => 'z'.

Comment (penghuo) [reply core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/datetime/DateTimeFormatterUtil.java:121]: Got it, does it covered in the test cases also?

Comment (lyndonb-bq) [reply core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/datetime/DateTimeFormatterUtil.java:121]: Yes

**Review by penghuo — APPROVED** [review]
Thanks for the change!

**Pull request merged**